### PR TITLE
[5.3] Fire BelongsToMany relation events for attach/detach/sync/toggle

### PR DESCRIPTION
**Pull request by jarektkaczyk**
BelongsToMany relations is the only missing part in the **events** system in eloquent.

It has been requested for a long time and has many applications, eg. 
#10435 #10545 #10770 #4022 #2303 #1703 #8452

_This **PR is BC** however it also brings change in the public interface, that is all the methods (`attach/detach/sync/toggle`) get additional `false` return value in case `*ing` event halts the execution. As is it doesn't break BC however there might be edge case:_

_1. one relies on `array` return value from `sync` or `toggle`_
_2. then adds `syncing/toggling` event handler that can halt execution in some circumstances_
_3. the return value under those circumstances is no longer `array` but `false`_

---
- It introduces 8 new events:

``` php
eloquent.attaching.[RELATION]: Parent\Model
eloquent.attached.[RELATION]: Parent\Model
eloquent.detaching._RELATION_: Parent\Model
eloquent.detached._RELATION_: Parent\Model
eloquent.syncing._RELATION_: Parent\Model
eloquent.synced._RELATION_: Parent\Model
eloquent.toggling._RELATION_: Parent\Model
eloquent.toggled._RELATION_: Parent\Model
```
- and appropriate methods on the `Model` to register these events (due to different nature, they have additional parameter - relation name), eg:

``` php
// Model::attaching($relation, function ($parent, Support\Collection $pivotData) { ... })
User::attaching('roles', function ($user, $pivotData) {
    // payload contains parent of the relation AND pivot data being inserted to the table
    $pivotData == Collection([
        ['user_id' => $user->id, 'role_id' => ATTACHED_ID, 'some_pivot_field' => 'some pivot value'],
        ...
    ]);
})

// manually
$dispatcher->listen('eloquent.attaching.roles: App\User', function ($user, $pivotData) { ... })
```

Since the `$pivotData` payload may vary depending on the event, here's complete list of its content:
- `attaching` - like in the example above - Collection with **raw data to be inserted** to the pivot table
- `attached` - same as above
- `detaching` - Collection of **ids** being detached OR empty Collection
- `detaching` - Collection of detached **ids** OR empty Collection
- `syncing` - Collection of **id => [ ..pivot data.. ]** entries to be inserted/updated
- `synced` - Collection of **changes** made (the same as return value of `sync` only wrapped in Collection)
- `toggling` - Collection of **id => [ ..pivot data.. ]** entries to be inserted/updated
- `toggled` - Collection of **changes** made (the same as return value of `sync` only wrapped in Collection)

and an example of all events being fired when `syncing`:

``` php
>>> $user->roles()->sync([1, 2 => ['extra' => 'provided pivot value']])

user 1 syncing roles: {"1":[],"2":{"extra":"provided pivot value"}}
user 1 detaching roles: ["6"]
user 1 detached roles: ["6"]
user 1 attaching roles: [{"user_id":1,"role_id":1}]
user 1 attached roles: [{"user_id":1,"role_id":1,}]
user 1 attaching roles: [{"user_id":1,"role_id":2,"extra":"provided pivot value"}]
user 1 attached roles: [{"user_id":1,"role_id":2,"extra":"provided pivot value"}]
user 1 synced roles: {"attached":[1,2],"detached":[6],"updated":[]}
```

---

All of the `*ing` events are, just like existing ones, capable of **halting** the action if your listener returns `false`:

``` php
// Bill can't be an admin:
User::attaching('roles', function ($user, $pivotData) use ($adminRole) {
   if ($user->name == 'Bill'
      && $pivotData->contains(function ($record) use ($adminRole) {
           return $record['role_id'] == $adminRole->id;
      })
   ) {
       return false;
   }
})
```

You can also adjust pivot data in the callback, just like you'd do with exiting events:

``` php
// We need UUID for the pivot records (well... yeah):
User::attaching('roles', function ($user, $pivotData) {
   foreach ($pivotData as $key => $record) {
      $record['uuid'] = data_get($record, 'uuid', Uuid::generate()); // generate if not provided already
      $pivodData->put($key, $record);
   } 
})


>>> $user->roles()->attach(5)
>>> $user->roles()->find(5)
Role {
  'name' => 'editor',
  ...
  'pivot' => Pivot {
    'user_id' => 1,
    'role_id' => 5,
    'uuid' => 'a-nice-auto-generated-uuid'
  }
}
```
### Notes on the commits:
1. the code in `BelongsToMany` has been a bit refactored in order to use `Collection` where events require that
2. new tests added + some brittle tests refactored

### CODE REVIEW

**Comment by axyr** [comment]
I really like this.

Should $model->belongsToRelation()->**associate**($otherModel)

not also have an event?

**Comment by jarektkaczyk** [comment]
1 this is only for belongsToMany, that is operations on pivot table, which
can't be tracked on any way in eloquent directly, unlike belongsTo
2 associate doesn't run the query, so it's not real association until you
call save

Jarek Tkaczyk

softonsofa :: development with pleasure

File: tests/Database/DatabaseEloquentBelongsToManyTest.php [hunk 1]
```diff
@@ -609,53 +608,162 @@ public function testTouchIfTouching()
...
     {
-        $relation = $this->getMockBuilder('Illuminate\Database\Eloquent\Relations\BelongsToMany')->setMethods(['attach', 'detach', 'touchIfTouching', 'formatRecordsList'])->setConstructorArgs($this->getRelationArguments())->getMock();
+        $relation = $this->getMockBuilder('Illuminate\Database\Eloquent\Relations\BelongsToMany')->setMethods(['attach', 'detach', 'touchIfTouching'])->setConstructorArgs($this->getRelationArguments())->getMock();
+        $relation->getQuery()->shouldReceive('where')->once()->with('user_role.foo', '=', 'bar', 'and')->andReturnSelf();
+        $currentQuery = $this->getNewQuery($relation);
+        $currentQuery->shouldReceive('pluck')->andReturn(new BaseCollection([1, 2, 3]));
 
-        // we expect to call $relation->wherePivot()
-        $relation->getQuery()->shouldReceive('where')->once()->andReturn($relation);
+        // This is our test! The wherePivot() params also need to be called
```
Comment (taylorotwell) [inline tests/Database/DatabaseEloquentBelongsToManyTest.php]: What does "This is our test!" mean?

Comment (jarektkaczyk) [inline tests/Database/DatabaseEloquentBelongsToManyTest.php]: We should ask @GrahamCampbell :)
I guess it refers to the `wherePivot` line below. There are bunch of tests in this class that are mocking CUT and require refactoring, but I didn't want to do that, so PR doesn't get even bigger and harder to review.

**Comment by jarektkaczyk** [comment]
ping @taylorotwell

**Comment by GrahamCampbell** [comment]
> ping @taylorotwell

Pending you removing "This is our test!".

**Comment by jarektkaczyk** [comment]
WAT?

That's an old test, out of scope of this PR, and btw your test Graham (or at least git blames you) ;)

**Comment by GrahamCampbell** [comment]
It git bames to you clearly?

**Comment by GrahamCampbell** [comment]
image

**Comment by jarektkaczyk** [comment]
well, as irrelevant as it is, here's the time travel: https://github.com/laravel/framework/blob/d70921d0e22f4f0b6da51475ab111ce67f3ed383/tests/Database/DatabaseEloquentBelongsToManyTest.php#L373

so in the git blamed your `tabs to spaces` as latest.

Now, back to the feature being proposed.

**Comment by GrahamCampbell** [comment]
I'd appreciate you not taking the piss. Just remove it. I don't care who added it.

**Comment by taylorotwell** [comment]
I really don't think this is ever a good solution to a problem to be honest. If you want something to happen before or after you attach / detach / toggle / whatever, you should make a repository class that actually does those things before calling ->attach. That saves us having to maintain all of this complicated event code throughout Eloquent.

**Comment by andrewkamm** [comment]
@jarektkaczyk Did you find a clean way to implement this since this PR was rejected? I agree that this would be really useful to have in Laravel core.

**Comment by jarektkaczyk** [comment]
I didn't work on it anymore since then.

Jarek Tkaczyk

softonsofa :: development with pleasure

**Comment by lukepolo** [comment]
Just manually fire the events , is what I did

**Comment by andyberry88** [comment]
We needed attach/detach events at [Agency Core](https://agencycore.net/) for sending things like User notifications when new tasks were assigned.

I've created a Gist with the extensions we added to make it work - https://gist.github.com/andyberry88/be3c45380568fc359cb61e00c4249704. It's basically the changes suggested by @jarektkaczyk above but taken out into app classes. I've had to extract it from other extensions we've added so hopefully it works. It should at least give an idea of the changes required.

**Comment by andrewkamm** [comment]
@andyberry88 Thanks!

**Comment by lukepolo** [comment]
@jarektkaczyk maybe we could work together to make it a package ?  Not sure if theres a ton of demand for it but it would save me time in some other applications.

**Comment by jarektkaczyk** [comment]
@lukepolo I can help you out with the package if you need, just drop me an email with what assistance you need :surfing_man:

**Comment by lephleg** [comment]
I can also assist any attempt to make this formal. 
I am really having a hard time to find a proper workaround for this, as I need to track relationship changes for logging purposes in L4.

**Comment by lukepolo** [comment]
I sadly do not have the time todo this as previously thought, you can see if @jarektkaczyk still wants to help out

**Comment by neylsongularte** [comment]
Experimental package:

https://github.com/neylsongularte/eloquent-extra-events

**Comment by fico7489** [comment]
I created a well documented and well-tested package for this problem  https://github.com/fico7489/laravel-pivot.  
You can do much more with my package that with this one neylsongularte/eloquent-extra-events.
You can get parent model,  changed pivot ids, and on which relation is updating performed.

**Comment by jehoshua02** [comment]
@taylorotwell If we create a repository class, is there a way to get Nova to utilize it?

We are doing some caching and want to invalidate/recalculate the cache when many-to-many relationships change.

Do you have any recommendations for how we can do this? For now I will see if we can use laravel-pivot.

**Comment by vpillinger** [comment]
@taylorotwell Sorry to necro this comment on you; however, perhaps there is a better solution here than creating new events or leaving the current functionality as-is?

```
If you want something to happen before or after you attach / detach / toggle / whatever, you should make a repository class that actually does those things before calling ->attach.
```

While I generally agree with this statement regarding business logic like sending notifications, there are good use-cases for wanting events when updating pivots. For example, a common use-case is to implement auditing and having some sort of events fire allows one to setup auditing events in a general way. For example, in an auditing package.

My understanding is that sync() does not trigger the normal eloquent events even if we create a custom pivot model. Therefore, currently one can audit on a belongsTo relationship by listening to the model being updated, but there is not a way to listen to a belongsToMany relationship as the information is contained in an unrelated model that isn't triggering events. 

So maybe the appropriate solution is to just make sync() and related methods trigger these default events if a custom Pivot model exists?

The current workaround is to just ignore these relationship functions and update the pivots manually with a save() statement to get eloquent events.

**Pull request closed without merging**